components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+// Package oneaxis implements a one-axis gantry.
+package oneaxis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"go.uber.org/multierr"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/gantry"
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	rdkutils "go.viam.com/rdk/utils"
+)
+
+var modelname = resource.NewDefaultModel("oneaxis")
+
+func errBadNumLimitSwitches(name string, numLimPins int) error {
+	return fmt.Errorf("bad number of limit switch pins for gantry %s: can only be 0, 1 or 2, have %d",
+		name, numLimPins)
+}
+
+func errDimensionsNotFound(name string, length, mmPerRev float64) error {
+	return fmt.Errorf("zero dimension found for gantry %s, length: %.2f, mmPerRev %.2f",
+		name, length, mmPerRev)
+}
+
+var errZeroLengthGantry = errors.New("zero length gantry found between position limits")
+
+// AttrConfig is used for converting oneAxis config attributes.
+type AttrConfig struct {
+	Board           string    `json:"board,omitempty"` // used to read limit switch pins and control motor with gpio pins
+	Motor           string    `json:"motor"`
+	LimitSwitchPins []string  `json:"limit_pins,omitempty"`
+	LimitPinEnabled *bool     `json:"limit_pin_enabled_high,omitempty"`
+	LengthMm        float64   `json:"length_mm"`
+	MmPerRevolution float64   `json:"mm_per_rev,omitempty"`
+	GantryRPM       float64   `json:"gantry_rpm,omitempty"`
+	Axis            r3.Vector `json:"axis"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (cfg *AttrConfig) Validate(path string) ([]string, error) {
+	var deps []string
+
+	if len(cfg.Motor) == 0 {
+		return nil, errors.New("cannot find motor for gantry")
+	}
+	deps = append(deps, cfg.Motor)
+
+	if cfg.LengthMm <= 0 {
+		return nil, errors.New("each axis needs a non-zero and positive length")
+	}
+
+	if len(cfg.LimitSwitchPins) == 0 && len(cfg.Board) > 0 {
+		return nil, errors.New("gantry with encoders have to assign boards or controllers to motors")
+	}
+
+	if len(cfg.Board) == 0 && len(cfg.LimitSwitchPins) > 0 {
+		return nil, errors.New("cannot find board for gantry")
```

Comment:
could we include a little more detail in this error message? since you only need to supply a board if you're using a limit switch gantry, this message could be misleading to a user (I like the specificity you used in L. 65, for example)